src/daemon/ca/rc.rs

```diff
@@ -450,18 +450,42 @@ impl ResourceClass {
     }
 
     /// Activate a new key, if it's been longer than the staging period.
-    pub fn keyroll_activate(&self, staging_time: Duration, signer: &KrillSigner) -> KrillResult<Vec<CaEvtDet>> {
-        if !self.key_state.has_new_key()
-            || (staging_time > Duration::seconds(0) && self.last_key_change + staging_time > Time::now())
-        {
-            return Ok(vec![]);
-        }
+    pub fn keyroll_activate(
+        &self,
+        staging_time: Duration,
+        issuance_timing: &IssuanceTimingConfig,
+        signer: &KrillSigner,
+    ) -> KrillResult<Vec<CaEvtDet>> {
+        if let Some(new_key) = self.key_state.new_key() {
+            if staging_time > Duration::seconds(0) && self.last_key_change + staging_time > Time::now() {
+                Ok(vec![])
+            } else {
+                let key_activated =
+                    self.key_state
+                        .keyroll_activate(self.name.clone(), self.parent_rc_name.clone(), signer)?;
+
+                let roa_updates = self.roas.activate_key(new_key, issuance_timing, signer)?;
+                let roas_updated = CaEvtDet::RoasUpdated {
+                    resource_class_name: self.name.clone(),
+                    updates: roa_updates,
+                };
 
-        Ok(vec![self.key_state.keyroll_activate(
-            self.name.clone(),
-            self.parent_rc_name.clone(),
-            signer,
-        )?])
+                let mut cert_updates = ChildCertificateUpdates::default();
+                for issued in self.certificates.iter() {
+                    // re-issue
+                    let re_issued = self.re_issue(issued, None, new_key, None, issuance_timing, signer)?;
```

Comment:
If `self.re_issue()` succeeds on one loop iteration but fails on the next this fn will abort, but in that case the event for the update that happened will not be returned. Is this okay?

Reply:
That is okay, none of this modifies the actual CA. This is creating a number of events which will only be applied if everything was successful - including triggered pre-save effects (create new mft and crl in particular).